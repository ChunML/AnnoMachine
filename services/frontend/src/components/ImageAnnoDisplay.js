import React, { useState } from 'react';
import PropTypes from 'prop-types';

function ImageAnnoDisplay(props) {
  const {
    svgWidth,
    svgHeight,
    imageWidth,
    imageHeight,
    drawBoxes,
    scale,
    name,
    onImageClick,
    createMessage,
  } = props;

  const [moveable, setMoveable] = useState(false);

  const handleMouseMove = e => {
    if (!moveable) return;
    if (drawBoxes.length > 1) {
      createMessage(
        'error',
        'Only one box can be editted by clicking at a time.'
      );
      return;
    }
    if (drawBoxes.length === 0) {
      return;
    }
    const rect = e.target.getBoundingClientRect();
    const clickX = (e.clientX - rect.x) / scale;
    const clickY = (e.clientY - rect.y) / scale;
    const newBox = { ...drawBoxes[0] };
    if (
      Math.abs(clickX - drawBoxes[0].x_min) >
      Math.abs(clickX - drawBoxes[0].x_max)
    ) {
      newBox.x_max = clickX;
    } else {
      newBox.x_min = clickX;
    }
    if (
      Math.abs(clickY - drawBoxes[0].y_min) >
      Math.abs(clickY - drawBoxes[0].y_max)
    ) {
      newBox.y_max = clickY;
    } else {
      newBox.y_min = clickY;
    }
    onImageClick(newBox);
  };

  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
      className="image-large"
      width={svgWidth}
      height={svgHeight}
      onClick={() => setMoveable(!moveable)}
    >
      <image
        xlinkHref={`${process.env.REACT_APP_API_URL}/api/uploads/${name}`}
        x="0"
        y="0"
        width={svgWidth}
        height={svgHeight}
        style={
          imageWidth > imageHeight ? { width: '100%' } : { height: '100%' }
        }
        onMouseMove={handleMouseMove}
      />
      {drawBoxes.length > 0 &&
        drawBoxes.map(box => (
          <rect
            key={box.id}
            x={box.x_min * scale}
            y={box.y_min * scale}
            width={(box.x_max - box.x_min) * scale}
            height={(box.y_max - box.y_min) * scale}
            style={{ fill: 'none', stroke: 'lime', strokeWidth: '3' }}
          />
        ))}
    </svg>
  );
}

ImageAnnoDisplay.propTypes = {
  svgWidth: PropTypes.number.isRequired,
  svgHeight: PropTypes.number.isRequired,
  imageWidth: PropTypes.number.isRequired,
  imageHeight: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  scale: PropTypes.number.isRequired,
  drawBoxes: PropTypes.array.isRequired,
  onImageClick: PropTypes.func.isRequired,
  createMessage: PropTypes.func.isRequired,
};

export default ImageAnnoDisplay;
