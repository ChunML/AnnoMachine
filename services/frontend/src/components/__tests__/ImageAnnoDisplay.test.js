import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import ImageAnnoDisplay from '../ImageAnnoDisplay';
import '../../setupTests';

const props_one_box = {
  svgWidth: 100,
  svgHeight: 100,
  drawBoxes: [
    {
      id: 1,
      x_min: 10,
      y_min: 10,
      x_max: 20,
      y_max: 20,
    },
  ],
  scale: 0.7,
  name: 'testName',
};

const props_two_boxes = {
  svgWidth: 100,
  svgHeight: 100,
  drawBoxes: [
    {
      id: 1,
      x_min: 10,
      y_min: 10,
      x_max: 20,
      y_max: 20,
    },
    {
      id: 2,
      x_min: 20,
      y_min: 20,
      x_max: 30,
      y_max: 30,
    },
  ],
  scale: 0.7,
  name: 'testName',
};

describe('Display landscape image with one box', () => {
  const imageSize = { imageWidth: 100, imageHeight: 50 };
  const onImageClick = jest.fn();
  const createMessage = jest.fn();
  const props = props_one_box;
  const component = (
    <ImageAnnoDisplay
      {...{ ...props, ...imageSize, onImageClick, createMessage }}
    />
  );
  it('ImageAnnoDisplay renders properly', () => {
    const wrapper = shallow(component);
    const svg = wrapper.find('svg');
    expect(svg.length).toBe(1);
    const image = svg.find('image');
    expect(image.length).toBe(1);
    expect(image.get(0).props.style.width).toEqual('100%');
    expect(onImageClick).toHaveBeenCalledTimes(0);
    expect(createMessage).toHaveBeenCalledTimes(0);
    svg.simulate('click');
    wrapper.find('image').simulate('mouseMove', {
      target: {
        getBoundingClientRect: () => ({ x: 100, y: 100 }),
      },
      clientX: 20,
      clientY: 20,
    });
    expect(onImageClick).toHaveBeenCalledTimes(1);
    expect(createMessage).toHaveBeenCalledTimes(0);
    const rects = svg.find('rect');
    expect(rects.length).toBe(1);
    rects.getElements().forEach((rect, i) => {
      expect(rect.props.x).toEqual(props.drawBoxes[i].x_min * props.scale);
      expect(rect.props.y).toEqual(props.drawBoxes[i].y_min * props.scale);
      expect(rect.props.width).toEqual(
        (props.drawBoxes[i].x_max - props.drawBoxes[i].x_min) * props.scale
      );
      expect(rect.props.height).toEqual(
        (props.drawBoxes[i].y_max - props.drawBoxes[i].y_min) * props.scale
      );
    });
  });

  it('ImageAnnoDisplay renders a snapshot properly', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Display landscape image with two boxes', () => {
  const imageSize = { imageWidth: 100, imageHeight: 50 };
  const onImageClick = jest.fn();
  const createMessage = jest.fn();
  const props = props_two_boxes;
  const component = (
    <ImageAnnoDisplay
      {...{ ...props, ...imageSize, onImageClick, createMessage }}
    />
  );
  it('ImageAnnoDisplay renders properly', () => {
    const wrapper = shallow(component);
    const svg = wrapper.find('svg');
    expect(svg.length).toBe(1);
    const image = svg.find('image');
    expect(image.length).toBe(1);
    expect(image.get(0).props.style.width).toEqual('100%');
    expect(onImageClick).toHaveBeenCalledTimes(0);
    expect(createMessage).toHaveBeenCalledTimes(0);
    svg.simulate('click');
    wrapper.find('image').simulate('mouseMove', {
      target: {
        getBoundingClientRect: () => ({ x: 100, y: 100 }),
      },
      clientX: 20,
      clientY: 20,
    });
    expect(onImageClick).toHaveBeenCalledTimes(0);
    expect(createMessage).toHaveBeenCalledTimes(1);
    const rects = svg.find('rect');
    expect(rects.length).toBe(2);
    rects.getElements().forEach((rect, i) => {
      expect(rect.props.x).toEqual(props.drawBoxes[i].x_min * props.scale);
      expect(rect.props.y).toEqual(props.drawBoxes[i].y_min * props.scale);
      expect(rect.props.width).toEqual(
        (props.drawBoxes[i].x_max - props.drawBoxes[i].x_min) * props.scale
      );
      expect(rect.props.height).toEqual(
        (props.drawBoxes[i].y_max - props.drawBoxes[i].y_min) * props.scale
      );
    });
  });

  it('ImageAnnoDisplay renders a snapshot properly', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Display portrait image', () => {
  const imageSize = { imageWidth: 50, imageHeight: 100 };
  const onImageClick = jest.fn();
  const createMessage = jest.fn();
  const props = props_one_box;
  const component = (
    <ImageAnnoDisplay
      {...{ ...props, ...imageSize, onImageClick, createMessage }}
    />
  );
  it('ImageAnnoDisplay renders properly', () => {
    const wrapper = shallow(component);
    const svg = wrapper.find('svg');
    expect(svg.length).toBe(1);
    const image = svg.find('image');
    expect(image.length).toBe(1);
    expect(image.get(0).props.style.height).toEqual('100%');
    expect(onImageClick).toHaveBeenCalledTimes(0);
    expect(createMessage).toHaveBeenCalledTimes(0);
    svg.simulate('click');
    wrapper.find('image').simulate('mouseMove', {
      target: {
        getBoundingClientRect: () => ({ x: 100, y: 100 }),
      },
      clientX: 20,
      clientY: 20,
    });
    expect(onImageClick).toHaveBeenCalledTimes(1);
    expect(createMessage).toHaveBeenCalledTimes(0);
    const rects = svg.find('rect');
    expect(rects.length).toBe(1);
    rects.getElements().forEach((rect, i) => {
      expect(rect.props.x).toEqual(props.drawBoxes[i].x_min * props.scale);
      expect(rect.props.y).toEqual(props.drawBoxes[i].y_min * props.scale);
      expect(rect.props.width).toEqual(
        (props.drawBoxes[i].x_max - props.drawBoxes[i].x_min) * props.scale
      );
      expect(rect.props.height).toEqual(
        (props.drawBoxes[i].y_max - props.drawBoxes[i].y_min) * props.scale
      );
    });
  });

  it('ImageAnnoDisplay renders a snapshot properly', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
