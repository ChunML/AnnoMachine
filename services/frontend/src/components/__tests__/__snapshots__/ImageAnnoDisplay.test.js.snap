// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display landscape image ImageAnnoDisplay renders a snapshot properly 1`] = `
<svg
  height={100}
  style={
    Object {
      "marginLeft": "-14px",
    }
  }
  width={100}
  xmlns="http://www.w3.org/2000/svg"
  xmlnsXlink="http://www.w3.org/1999/xlink"
>
  <image
    style={
      Object {
        "width": "100%",
      }
    }
    x="0"
    xlinkHref="http://localhost/api/uploads/testName"
    y="0"
  />
  <rect
    height={7}
    style={
      Object {
        "fill": "none",
        "stroke": "lime",
        "strokeWidth": "3",
      }
    }
    width={7}
    x={7}
    y={7}
  />
  <rect
    height={7}
    style={
      Object {
        "fill": "none",
        "stroke": "lime",
        "strokeWidth": "3",
      }
    }
    width={7}
    x={14}
    y={14}
  />
</svg>
`;

exports[`Display portrait image ImageAnnoDisplay renders a snapshot properly 1`] = `
<svg
  height={100}
  style={
    Object {
      "marginLeft": "-14px",
    }
  }
  width={100}
  xmlns="http://www.w3.org/2000/svg"
  xmlnsXlink="http://www.w3.org/1999/xlink"
>
  <image
    style={
      Object {
        "height": "100%",
      }
    }
    x="0"
    xlinkHref="http://localhost/api/uploads/testName"
    y="0"
  />
  <rect
    height={7}
    style={
      Object {
        "fill": "none",
        "stroke": "lime",
        "strokeWidth": "3",
      }
    }
    width={7}
    x={7}
    y={7}
  />
  <rect
    height={7}
    style={
      Object {
        "fill": "none",
        "stroke": "lime",
        "strokeWidth": "3",
      }
    }
    width={7}
    x={14}
    y={14}
  />
</svg>
`;
